openedx/core/djangoapps/schedules/tasks.py

```diff
@@ -56,14 +59,33 @@ def _recurring_nudge_schedule_send(site_id, msg_str):
 
 
 def _recurring_nudge_schedules_for_hour(target_hour, org_list, exclude_orgs=False):
+    beginning_of_day = target_hour.replace(hour=0, minute=0, second=0)
+    users = User.objects.filter(
+        courseenrollment__schedule__start__gte=beginning_of_day,
+        courseenrollment__schedule__start__lt=beginning_of_day + datetime.timedelta(days=1),
+        courseenrollment__is_active=True,
+    ).annotate(
+        first_schedule=Min('courseenrollment__schedule__start')
+    ).filter(
+        first_schedule__gte=target_hour,
+        first_schedule__lt=target_hour + datetime.timedelta(minutes=60)
+    )
+
+    if org_list is not None:
+        if exclude_orgs:
+            users = users.exclude(courseenrollment__course__org__in=org_list)
+        else:
+            users = users.filter(courseenrollment__course__org__in=org_list)
+
     schedules = Schedule.objects.select_related(
         'enrollment__user__profile',
         'enrollment__course',
     ).filter(
-        start__gte=target_hour,
-        start__lt=target_hour + datetime.timedelta(minutes=60),
+        enrollment__user__id__in=users,
```

Comment:
Couldn't this potentially be thousands of user ids? Or does the ORM transparently handle issues like this...

Reply:
The ORM is not that smart. This will indeed result in a giant `WHERE` clause. We face a similar issue in edx/course-discovery when converting from IDs stored in Elasticsearch to model instances. (It turns out SQLite (used for dev) has a limit to the size of its queries!)

If you use `enrollment__user=users`, the ORM should generate a `JOIN`.